File: csharp/tests/Idkollen.Client.Tests/DocumentTests.cs
using System.Net;
using System.Text;
using Idkollen.Client.Models;
using Xunit;

namespace Idkollen.Client.Tests;

public class DocumentTests
{
    private static (Idkollen.Client.IdkollenClient Client, MockHttpMessageHandler Handler) Build()
    {
        var handler = new MockHttpMessageHandler();
        var http = new HttpClient(handler);
        var client = new Idkollen.Client.IdkollenClientBuilder("cid", "sec")
            .BaseUrl("https://x.test").HttpClient(http).Build();
        return (client, handler);
    }

    [Fact]
    public async Task Upload()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { id = "doc1", hash = "h1" });
        var data = Encoding.UTF8.GetBytes("PDF-DATA");
        var result = await client.Document.UploadAsync(data, "contract.pdf");
        Assert.Equal("doc1", result.Id);
        Assert.Equal("h1", result.Hash);

        var req = handler.LastRequest;
        Assert.Equal(HttpMethod.Post, req.Method);
        Assert.StartsWith("multipart/form-data; boundary=", req.Content!.Headers.ContentType!.ToString());
        Assert.Contains("contract.pdf", handler.LastRequestBody);
        Assert.Contains("PDF-DATA", handler.LastRequestBody);
    }

    [Fact]
    public async Task DownloadReturnsRawBytes()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, "\x25PDF-binary-bytes");
        var result = await client.Document.DownloadAsync("doc1");
        Assert.Equal("\x25PDF-binary-bytes", Encoding.UTF8.GetString(result));
    }

    [Fact]
    public async Task Delete()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.NoContent, "");
        await client.Document.DeleteAsync("doc1");
        Assert.Equal(HttpMethod.Delete, handler.LastRequest.Method);
        Assert.Equal("/document/doc1", handler.LastRequest.RequestUri!.AbsolutePath);
    }
}

File: swift/Tests/IdkollenClientTests/DocumentTests.swift
import XCTest
@testable import IdkollenClient

final class DocumentTests: XCTestCase {
    var client: IdkollenClient!

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
        client = IdkollenClientBuilder(clientId: "cid", clientSecret: "sec")
            .baseUrl("https://x.test")
            .urlSession(.mocked())
            .build()
    }

    func testUpload() async throws {
        MockURLProtocol.enqueue(status: 200, json: ["id": "doc1", "hash": "h1"])

        let pdfData = "PDF-DATA".data(using: .utf8)!
        let response = try await client.document.upload(data: pdfData, filename: "contract.pdf")

        XCTAssertEqual(response.id, "doc1")
        XCTAssertEqual(response.hash, "h1")

        let req = MockURLProtocol.capturedRequests.last!
        let contentType = req.value(forHTTPHeaderField: "Content-Type") ?? ""
        XCTAssertTrue(contentType.hasPrefix("multipart/form-data; boundary="), "Expected multipart content-type, got: \(contentType)")

        let bodyString = String(data: MockURLProtocol.capturedBodies.last!, encoding: .utf8) ?? ""
        XCTAssertTrue(bodyString.contains("filename=\"contract.pdf\""), "Body should contain filename")
        XCTAssertTrue(bodyString.contains("Content-Type: application/pdf"), "Body should contain mime type")
        XCTAssertTrue(bodyString.contains("PDF-DATA"), "Body should contain file data")
    }

    func testDownloadReturnsRawBytes() async throws {
        let expected = Data([0x25, 0x50, 0x44, 0x46]) + "binary".data(using: .utf8)!
        MockURLProtocol.enqueue(status: 200, body: expected)

        let result = try await client.document.download(id: "doc1")
        XCTAssertEqual(result, expected)
    }

    func testDelete() async throws {
        MockURLProtocol.enqueue(status: 204, body: Data())

        try await client.document.delete(id: "doc1")

        let req = MockURLProtocol.capturedRequests.last!
        XCTAssertEqual(req.httpMethod, "DELETE")
        XCTAssertEqual(req.url?.path, "/document/doc1")
    }
}
